### PR TITLE
git-init: add support for fetching submodules 🚝

### LINKED ISSUE #1523: tekton pipeline git resource should init and update submodules recursively.
# Expected Behavior

Suppose we have a git pipeline resource as described in the doc:
https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#git-resource

```
apiVersion: tekton.dev/v1alpha1
kind: PipelineResource
metadata:
  name: wizzbang-git
  namespace: default
spec:
  type: git
  params:
    - name: url
      value: https://github.com/wizzbangcorp/wizzbang.git
    - name: revision
      value: master
```

Suppose that `https://github.com/wizzbangcorp/wizzbang.git` has `.gitmodules` with one or more git submodules.

When the git resource is cloned, the behavior should be equivalent to this:

```
git init <dir>
cd <dir>
git remote add origin <remote url>
git fetch --depth=1 --recurse-submodules=yes origin <revision>
git submodule init
git submodule update --recursive
```

where `<remote url>` would be `https://github.com/wizzbangcorp/wizzbang.git`

# Actual Behavior

Based on the latest commit of `pkg/git/git.go` here: https://github.com/tektoncd/pipeline/blob/ea94852c3aa046d9c898cd63eabc3440a4bf63b3/pkg/git/git.go#L47

This is equivalent to the following:

```
git init <dir>
cd <dir>
git remote add origin <remote url>
git fetch --depth=1 --recurse-submodules=yes origin <revision>
```

In other words, the submodules have not been initialized nor updated recursively.

# Steps to Reproduce the Problem

1. Execute a pipeline or task run that depends on submodule resources.

# Additional Info

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

It will do those commands by default as they are no-op.

Closes #1523 

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
`git-init` (thus `GitResource`) will now init and update submodules recursively by default
```

### CODE REVIEW

File: pkg/git/git.go [hunk 1]
```diff
@@ -111,3 +111,45 @@ func Commit(logger *zap.SugaredLogger, revision, path string) (string, error) {
 	}
 	return strings.TrimSuffix(output, "\n"), nil
 }
+
+func SubmoduleFetch(logger *zap.SugaredLogger, path string) error {
+	// HACK: This is to get git+ssh to work since ssh doesn't respect the HOME
```
Comment (sbwsg) [inline pkg/git/git.go]: This looks like this hack is copied from `Fetch()`.  Opportunity to de-dup by creating a helper func?

Comment (vdemeester) [reply pkg/git/git.go]: @sbwsg indeed :stuck_out_tongue: I was a bit lazy 🤫, I need to enhance that :wink:

File: cmd/git-init/main.go [hunk 2]
```diff
@@ -29,6 +29,7 @@ var (
 	revision               = flag.String("revision", "", "The Git revision to make the repository HEAD")
 	path                   = flag.String("path", "", "Path of directory under which git repository will be copied")
 	terminationMessagePath = flag.String("terminationMessagePath", "/dev/termination-log", "Location of file containing termination message")
+	submodules             = flag.Bool("submodules", true, "Initialize and fetch git submodules")
```
Comment (ImJasonH) [inline cmd/git-init/main.go]: WDYT about surfacing this option all the way up into the Git resource, so users can turn it off if they'd prefer?

Comment (vdemeester) [reply cmd/git-init/main.go]: @ImJasonH I was waiting for someone to comment this to do it :stuck_out_tongue:

File: docs/resources.md [hunk 3]
```diff
@@ -252,6 +252,9 @@ Params that can be added are the following:
     (branch, tag, commit SHA or ref) to clone. You can use this to control what
     commit [or branch](#using-a-branch) is used. _If no revision is specified,
     the resource will default to `latest` from `master`._
+1.  `submodules`: defines if the resource should initialize and
+    fetch the submodules, value is either `true` or `false`. _If note
```
Comment (sbwsg) [inline docs/resources.md]: s/note/not/

File: pkg/git/git.go (hunk 1 above)
Comment (sbwsg) [reply pkg/git/git.go]: nice :+1:

**Comment by NicolasRouquette** [comment]
As a new tekton fanboy here at JPL, I'm impressed by the quality of the PR & discussion. 
I look forward to using this update here.

**Pull request merged**